File: test_phase2_complete.py
"""
Phase 2 Complete Testing Suite - Citation Navigation & Enhanced Sidebar
"""

import requests
import json
import asyncio
from typing import Dict, Any

class Phase2CompleteTester:
    def __init__(self, base_url: str = "http://localhost:8080"):
        self.base_url = base_url
        self.test_user = "phase2_complete_test_user"
        self.results = []
        
    def log_result(self, test_name: str, success: bool, details: str = "", data: Any = None):
        """Log test result"""
        result = {
            "test": test_name,
            "success": success,
            "details": details,
            "data": data
        }
        self.results.append(result)
        
        status = "✅" if success else "❌"
        print(f"{status} {test_name}: {details}")
    
    def test_citation_service_functionality(self):
        """Test citation service core functionality"""
        print(f"\n🔧 Testing Citation Service Core Functionality")
        
        try:
            # Test that citation service can be imported and initialized
            from services.citation_service import get_citation_service, CitationService
            
            service = CitationService()
            cache_stats = service.get_cache_stats()
            
            if isinstance(cache_stats, dict) and "total_entries" in cache_stats:
                self.log_result("Citation Service Import", True, "Service imported and initialized successfully")
                self.log_result("Citation Service Cache", True, f"Cache stats: {cache_stats}")
                
                # Test parsing functionality
                mock_data = {
                    'id': 'https://openalex.org/W123',
                    'doi': 'https://doi.org/10.1038/test',
                    'title': 'Test Paper',
                    'publication_year': 2023,
                    'cited_by_count': 50,
                    'ids': {'pmid': 'https://pubmed.ncbi.nlm.nih.gov/12345'},
                    'authorships': [{'author': {'display_name': 'Test Author'}}],
                    'host_venue': {'display_name': 'Test Journal'},
                    'referenced_works': [],
                    'abstract': 'Test abstract'
                }
                
                result = service._parse_openalex_response(mock_data)
                if result and result.title == 'Test Paper':
                    self.log_result("Citation Service Parsing", True, f"Parsed: {result.title} by {result.authors[0]}")
                else:
                    self.log_result("Citation Service Parsing", False, "Failed to parse mock data")
                    
            else:
                self.log_result("Citation Service Cache", False, "Invalid cache stats format")
                
        except Exception as e:
            self.log_result("Citation Service Functionality", False, f"Error: {str(e)}")
    
    def test_all_citation_endpoints(self):
        """Test all citation navigation endpoints"""
        print(f"\n🌐 Testing All Citation Navigation Endpoints")
        
        test_pmid = "test001"
        endpoints = [
            ("references", "📚 References"),
            ("references-network", "🕸️ References Network"),
            ("citations", "📈 Citations"),
            ("citations-network", "🌐 Citations Network")
        ]
        
        for endpoint, description in endpoints:
            try:
                response = requests.get(
                    f"{self.base_url}/articles/{test_pmid}/{endpoint}",
                    headers={"User-ID": self.test_user},
                    params={"limit": 10},
                    timeout=15
                )
                
                if response.status_code == 200:
                    data = response.json()
                    
                    if "network" in endpoint:
                        # Network endpoint
                        required_fields = ["nodes", "edges", "metadata"]
                        missing_fields = [field for field in required_fields if field not in data]
                        
                        if not missing_fields:
                            metadata = data["metadata"]
                            self.log_result(f"{description} Endpoint", True, 
                                          f"Network: {metadata['total_nodes']} nodes, {metadata['total_edges']} edges")
                        else:
                            self.log_result(f"{description} Endpoint", False, 
                                          f"Missing fields: {missing_fields}")
                    else:
                        # Data endpoint
                        required_fields = ["base_article", endpoint.replace('-', '_'), "total_found", "search_parameters", "cache_stats"]
                        missing_fields = [field for field in required_fields if field not in data]
                        
                        if not missing_fields:
                            self.log_result(f"{description} Endpoint", True, 
                                          f"Found {data['total_found']} {endpoint}")
                        else:
                            self.log_result(f"{description} Endpoint", False, 
                                          f"Missing fields: {missing_fields}")
                else:
                    self.log_result(f"{description} Endpoint", False, 
                                  f"HTTP {response.status_code}: {response.text[:100]}")
                    
            except Exception as e:
                self.log_result(f"{description} Endpoint", False, f"Error: {str(e)}")
    
    def test_frontend_api_proxies(self):
        """Test frontend API proxy routes"""
        print(f"\n🔗 Testing Frontend API Proxy Routes")
        
        # Note: This would require the frontend server to be running
        # For now, we'll just verify the files exist
        
        proxy_files = [
            "frontend/src/app/api/proxy/articles/[pmid]/references/route.ts",
            "frontend/src/app/api/proxy/articles/[pmid]/references-network/route.ts",
            "frontend/src/app/api/proxy/articles/[pmid]/citations/route.ts",
            "frontend/src/app/api/proxy/articles/[pmid]/citations-network/route.ts"
        ]
        
        import os
        for proxy_file in proxy_files:
            if os.path.exists(proxy_file):
                self.log_result(f"Proxy Route {proxy_file.split('/')[-2]}", True, "File exists")
            else:
                self.log_result(f"Proxy Route {proxy_file.split('/')[-2]}", False, "File missing")
    
    def test_enhanced_sidebar_component(self):
        """Test enhanced NetworkSidebar component"""
        print(f"\n📱 Testing Enhanced NetworkSidebar Component")
        
        import os
        sidebar_file = "frontend/src/components/NetworkSidebar.tsx"
        
        if os.path.exists(sidebar_file):
            self.log_result("NetworkSidebar Component", True, "Component file exists")
            
            # Check for key functionality in the file
            with open(sidebar_file, 'r') as f:
                content = f.read()
                
                key_features = [
                    ("Navigation Change Handler", "onNavigationChange"),
                    ("Collection Integration", "handleAddToCollection"),
                    ("References Display", "references.length"),
                    ("Citations Display", "citations.length"),
                    ("Paper Details", "data.title"),
                    ("Collection Selection", "selectedCollection")
                ]
                
                for feature_name, search_term in key_features:
                    if search_term in content:
                        self.log_result(f"Sidebar {feature_name}", True, "Feature implemented")
                    else:
                        self.log_result(f"Sidebar {feature_name}", False, "Feature missing")
        else:
            self.log_result("NetworkSidebar Component", False, "Component file missing")
    
    def test_networkview_integration(self):
        """Test NetworkView integration with enhanced sidebar"""
        print(f"\n🔗 Testing NetworkView Integration")
        
        import os
        networkview_file = "frontend/src/components/NetworkView.tsx"
        
        if os.path.exists(networkview_file):
            with open(networkview_file, 'r') as f:
                content = f.read()
                
                integration_features = [
                    ("Sidebar Import", "import NetworkSidebar"),
                    ("Sidebar State", "showSidebar"),
                    ("Collections State", "collections"),
                    ("Sidebar Handlers", "handleSidebarNavigationChange"),
                    ("Sidebar JSX", "<NetworkSidebar")
                ]
                
                for feature_name, search_term in integration_features:
                    if search_term in content:
                        self.log_result(f"NetworkView {feature_name}", True, "Integration implemented")
                    else:
                        self.log_result(f"NetworkView {feature_name}", False, "Integration missing")
        else:
            self.log_result("NetworkView Integration", False, "NetworkView file missing")
    
    def test_database_schema(self):
        """Test database schema for citation relationships"""
        print(f"\n🗄️ Testing Database Schema")
        
        try:
            from database import ArticleCitation, AuthorCollaboration
            from sqlalchemy import inspect
            
            # Test ArticleCitation model
            citation_columns = [col.name for col in inspect(ArticleCitation).columns]
            required_citation_columns = ['id', 'citing_pmid', 'cited_pmid', 'citation_context', 'created_at']
            
            missing_citation_cols = [col for col in required_citation_columns if col not in citation_columns]
            if not missing_citation_cols:
                self.log_result("ArticleCitation Schema", True, f"All required columns present: {citation_columns}")
            else:
                self.log_result("ArticleCitation Schema", False, f"Missing columns: {missing_citation_cols}")
            
            # Test AuthorCollaboration model
            collab_columns = [col.name for col in inspect(AuthorCollaboration).columns]
            required_collab_columns = ['id', 'author1_name', 'author2_name', 'collaboration_count', 'created_at']
            
            missing_collab_cols = [col for col in required_collab_columns if col not in collab_columns]
            if not missing_collab_cols:
                self.log_result("AuthorCollaboration Schema", True, f"All required columns present: {collab_columns}")
            else:
                self.log_result("AuthorCollaboration Schema", False, f"Missing columns: {missing_collab_cols}")
                
        except Exception as e:
            self.log_result("Database Schema", False, f"Error: {str(e)}")
    
    def run_comprehensive_test(self):
        """Run comprehensive Phase 2 testing"""
        print("🧪 Phase 2 Complete Testing Suite - Citation Navigation & Enhanced Sidebar")
        print("=" * 80)
        
        # Run all test categories
        self.test_citation_service_functionality()
        self.test_all_citation_endpoints()
        self.test_frontend_api_proxies()
        self.test_enhanced_sidebar_component()
        self.test_networkview_integration()
        self.test_database_schema()
        
        # Summary
        print("\n📊 Phase 2 Complete Testing Summary")
        print("=" * 40)
        
        total_tests = len(self.results)
        passed_tests = sum(1 for r in self.results if r['success'])
        failed_tests = total_tests - passed_tests
        
        print(f"Total Tests: {total_tests}")
        print(f"✅ Passed: {passed_tests}")
        print(f"❌ Failed: {failed_tests}")
        print(f"Success Rate: {(passed_tests/total_tests)*100:.1f}%")
        
        if failed_tests > 0:
            print("\n❌ Failed Tests:")
            for result in self.results:
                if not result['success']:
                    print(f"  - {result['test']}: {result['details']}")
        
        # Phase 2 completion status
        critical_features = [
            "Citation Service Import",
            "References Endpoint",
            "Citations Endpoint", 
            "NetworkSidebar Component",
            "NetworkView Sidebar Import"
        ]
        
        critical_passed = sum(1 for r in self.results if r['success'] and r['test'] in critical_features)
        critical_total = len(critical_features)
        
        if critical_passed == critical_total:
            status = "✅ PHASE 2 COMPLETE - READY FOR PHASE 3"
        elif critical_passed >= critical_total * 0.8:
            status = "⚠️ PHASE 2 MOSTLY COMPLETE - MINOR FIXES NEEDED"
        else:
            status = "❌ PHASE 2 INCOMPLETE - MAJOR ISSUES NEED RESOLUTION"
        
        print(f"\n🎯 Phase 2 Status: {status}")
        print(f"Critical Features: {critical_passed}/{critical_total} passing")
        
        return failed_tests == 0

if __name__ == "__main__":
    tester = Phase2CompleteTester()
    success = tester.run_comprehensive_test()
    
    if success:
        print("\n🚀 Phase 2 Complete! Ready to proceed with Phase 3: Timeline Visualization.")
    else:
        print("\n⚠️ Some Phase 2 tests failed. Please review and fix issues before proceeding.")

File: test_phase2_complete.js
#!/usr/bin/env node

/**
 * PHASE 2 COMPLETION TEST
 * 
 * Comprehensive test to verify all Phase 2 fixes are working:
 * 1. PubMed Advanced Search Functionality
 * 2. Network Analysis Endpoints  
 * 3. Paper Analysis Tools
 */

const BACKEND_URL = 'https://r-dagent-production.up.railway.app';
const USER_ID = 'fredericle77@gmail.com';

async function testEndpoint(name, url, expectedKeys = []) {
    try {
        console.log(`🧪 Testing: ${name}`);
        
        const startTime = Date.now();
        const response = await fetch(url, {
            method: 'GET',
            headers: {
                'User-ID': USER_ID,
                'Content-Type': 'application/json'
            }
        });
        const responseTime = Date.now() - startTime;
        
        const responseText = await response.text();
        let responseData;
        try {
            responseData = JSON.parse(responseText);
        } catch (e) {
            responseData = { raw_response: responseText };
        }
        
        if (response.ok) {
            // Check for expected keys
            const missingKeys = expectedKeys.filter(key => !(key in responseData));
            if (missingKeys.length === 0) {
                console.log(`✅ ${name} - SUCCESS (${response.status}, ${responseTime}ms)`);
                
                // Log key metrics
                if (responseData.articles && responseData.articles.length) {
                    console.log(`   📄 Articles: ${responseData.articles.length}`);
                }
                if (responseData.topics && responseData.topics.length) {
                    console.log(`   🏷️  Topics: ${responseData.topics.length}`);
                }
                if (responseData.similarity_metrics) {
                    console.log(`   📊 Similarity: ${responseData.similarity_metrics.overall_similarity}`);
                }
                if (responseData.network && responseData.network.authors) {
                    console.log(`   👥 Authors: ${Object.keys(responseData.network.authors).length}`);
                }
                
                return { success: true, responseTime, data: responseData };
            } else {
                console.log(`❌ ${name} - MISSING KEYS: ${missingKeys.join(', ')}`);
                return { success: false, error: `Missing keys: ${missingKeys.join(', ')}` };
            }
        } else {
            console.log(`❌ ${name} - FAILED (${response.status}, ${responseTime}ms)`);
            if (responseData.detail) {
                console.log(`   Error: ${responseData.detail}`);
            }
            return { success: false, status: response.status, error: responseData.detail };
        }
        
    } catch (error) {
        console.log(`💥 ${name} - ERROR: ${error.message}`);
        return { success: false, error: error.message };
    }
}

async function runPhase2CompletionTest() {
    console.log('🚀 PHASE 2 COMPLETION TEST');
    console.log('=' .repeat(60));
    console.log(`Backend URL: ${BACKEND_URL}`);
    console.log(`User ID: ${USER_ID}`);
    console.log('');
    
    const results = [];
    
    // PHASE 2 FIX #1: PubMed Advanced Search
    console.log('📋 PHASE 2 FIX #1: PubMed Advanced Search');
    console.log('-'.repeat(50));
    
    const pubmedTests = [
        {
            name: 'PubMed Advanced Search - Diabetes',
            url: `${BACKEND_URL}/pubmed/advanced-search?q=diabetes&limit=3`,
            expectedKeys: ['query', 'articles', 'total_found', 'status']
        },
        {
            name: 'PubMed Advanced Search - Cancer',
            url: `${BACKEND_URL}/pubmed/advanced-search?q=cancer&limit=2`,
            expectedKeys: ['query', 'articles', 'total_found', 'status']
        }
    ];
    
    for (const test of pubmedTests) {
        const result = await testEndpoint(test.name, test.url, test.expectedKeys);
        results.push(result);
        await new Promise(resolve => setTimeout(resolve, 200));
    }
    
    console.log('');
    
    // PHASE 2 FIX #2: Network Analysis Endpoints
    console.log('📋 PHASE 2 FIX #2: Network Analysis Endpoints');
    console.log('-'.repeat(50));
    
    const networkTests = [
        {
            name: 'Article Citations',
            url: `${BACKEND_URL}/articles/37024129/citations?limit=5`,
            expectedKeys: ['source_article', 'citations', 'total_count']
        },
        {
            name: 'Article References',
            url: `${BACKEND_URL}/articles/37024129/references?limit=5`,
            expectedKeys: ['source_article', 'references', 'total_count']
        },
        {
            name: 'Citation Enrichment Status',
            url: `${BACKEND_URL}/articles/37024129/enrich-citations`,
            expectedKeys: ['source_article', 'citation_enrichment_status', 'available_actions']
        },
        {
            name: 'Author Network',
            url: `${BACKEND_URL}/articles/37024129/author-network?depth=2`,
            expectedKeys: ['source_article', 'network', 'suggested_authors']
        },
        {
            name: 'Similar Articles Network',
            url: `${BACKEND_URL}/articles/37024129/similar-network?limit=5`,
            expectedKeys: ['nodes', 'edges', 'metadata']
        }
    ];
    
    for (const test of networkTests) {
        const result = await testEndpoint(test.name, test.url, test.expectedKeys);
        results.push(result);
        await new Promise(resolve => setTimeout(resolve, 200));
    }
    
    console.log('');
    
    // PHASE 2 FIX #3: Paper Analysis Tools
    console.log('📋 PHASE 2 FIX #3: Paper Analysis Tools');
    console.log('-'.repeat(50));
    
    const analysisTests = [
        {
            name: 'Topic Analysis - Diabetes',
            url: `${BACKEND_URL}/papers/topic-analysis?q=diabetes&limit=5`,
            expectedKeys: ['query', 'papers_analyzed', 'topics', 'status']
        },
        {
            name: 'Topic Analysis - Machine Learning',
            url: `${BACKEND_URL}/papers/topic-analysis?q=machine%20learning&limit=3`,
            expectedKeys: ['query', 'papers_analyzed', 'topics', 'status']
        },
        {
            name: 'Similarity Analysis - Same Paper',
            url: `${BACKEND_URL}/papers/similarity-analysis?pmid1=37024129&pmid2=37024129`,
            expectedKeys: ['paper1', 'paper2', 'similarity_metrics', 'status']
        }
    ];
    
    for (const test of analysisTests) {
        const result = await testEndpoint(test.name, test.url, test.expectedKeys);
        results.push(result);
        await new Promise(resolve => setTimeout(resolve, 200));
    }
    
    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('📊 PHASE 2 COMPLETION TEST SUMMARY');
    console.log('='.repeat(60));
    
    const successful = results.filter(r => r.success);
    const failed = results.filter(r => !r.success);
    
    console.log(`✅ Successful: ${successful.length}/${results.length} (${(successful.length/results.length*100).toFixed(1)}%)`);
    console.log(`❌ Failed: ${failed.length}/${results.length} (${(failed.length/results.length*100).toFixed(1)}%)`);
    
    if (failed.length > 0) {
        console.log('\n❌ FAILED TESTS:');
        failed.forEach(result => {
            console.log(`   • ${result.name || 'Unknown'} - ${result.error || result.status}`);
        });
    }
    
    console.log('\n✅ SUCCESSFUL TESTS:');
    successful.forEach((result, index) => {
        const testName = pubmedTests.concat(networkTests, analysisTests)[index]?.name || `Test ${index + 1}`;
        console.log(`   • ${testName} (${result.responseTime}ms)`);
    });
    
    const overallSuccess = (successful.length / results.length) >= 0.9;
    console.log(`\n🎯 PHASE 2 STATUS: ${overallSuccess ? '✅ COMPLETE' : '❌ NEEDS WORK'}`);
    
    if (overallSuccess) {
        console.log('🎉 Phase 2 implementation is complete and working excellently!');
        console.log('');
        console.log('📋 PHASE 2 ACHIEVEMENTS:');
        console.log('   ✅ PubMed Advanced Search - Fully functional with structured results');
        console.log('   ✅ Network Analysis Endpoints - All endpoints working with proper data');
        console.log('   ✅ Paper Analysis Tools - Topic analysis and similarity comparison working');
        console.log('');
        console.log('🚀 Ready to proceed to Phase 3!');
    } else {
        console.log('🔧 Some Phase 2 features need additional work.');
    }
    
    return {
        total: results.length,
        successful: successful.length,
        failed: failed.length,
        successRate: successful.length / results.length,
        results: results
    };
}

// Run the test
runPhase2CompletionTest()
    .then(summary => {
        console.log(`\n📈 Final Phase 2 Success Rate: ${(summary.successRate * 100).toFixed(1)}%`);
        process.exit(summary.successRate >= 0.9 ? 0 : 1);
    })
    .catch(error => {
        console.error('💥 Phase 2 test execution failed:', error);
        process.exit(1);
    });
